mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,339 @@
+// limitations under the License.
+
+#define DEBUG_TYPE "splitmultipletapes"
+
+#include <cassert>
+#include <vector>
+
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/Support/Debug.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/IR/Types.h"
+#include "mlir/Pass/Pass.h"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+#define GEN_PASS_DEF_SPLITMULTIPLETAPESPASS
+#ifdef GEN_PASS_DEF_SPLITMULTIPLETAPESPASS
+#define SmallSetSize 8
+#endif
+#include "Catalyst/Transforms/Passes.h.inc"
+
+struct SplitMultipleTapesPass : public impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass> {
+    using impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass>::SplitMultipleTapesPassBase;
+
+    unsigned int countTapes(const func::FuncOp &func)
+    {
+        // Count the number of quantum.device operations in a function
+        unsigned int count = 0;
+        func->walk([&](catalyst::quantum::DeviceInitOp op) { count++; });
+        return count;
+    } // countTapes()
+
+    void collectOperationsForEachTape(const func::FuncOp &func,
+                                      SmallVector<std::vector<Operation *>> &OpsEachTape)
+    {
+        // During tracing, each tape starts with a qdevice_p primitive
+        // This means each tape starts with a quantum.device
+        // and ends with a quantum.device_release
+
+        // The structure of the FuncOp looks like the following:
+        // func.func @circuit(...) -> ... {...} {
+        // preprocessing
+        // quantum.device[...]
+        // tape 1
+        // quantum.device_release
+        // quantum.device[...]
+        // tape 2
+        // quantum.device_release
+        // ... more tapes
+        // post processing and return
+        // }
+
+        // This function parses the operations in the funcop and puts them into
+        // the container as
+        // OpsEachTape = [[preprocessing ops], [tape1 ops], [tape2 ops], ..., [postprocessing ops]]
+
+        // special: [0] for pre and [-1] for post processing
+        unsigned int curTape = 0;
+
+        func->walk([&](Operation *op) {
+            if (op == func) {
+                return; // don't visit the funcop itself
+            }
+            if (isa<catalyst::quantum::DeviceInitOp>(op)) {
+                curTape++;
+            }
+            OpsEachTape[curTape].push_back(op);
+            if ((isa<catalyst::quantum::DeviceReleaseOp>(op)) &&
+                (curTape == OpsEachTape.size() - 2)) {
+                // reached post processing
+                curTape++;
+            }
```

Comment:
done

We now forloop through the toplevel ops in the funcop's (only) region's blocks

Reply:
@paul0403 I was thinking about blocks and I still kind of disagree with that approach. That is because the blocks will use jumps and the jumps may impose a different order. I.e.

```
def foo():
   .start
   jump .end
   .middle
   return
   jump .middle
```